### PR TITLE
feat: add `valibot$` validator and fix types of `zod$` implementation

### PR DESCRIPTION
**Pull request by fabian-hiller**
# Overview

Adds Valibot support with a `valibot$` validator similarly to how Zod is supported with `zod$`.

Related #4998, Credits @brandonpittman

# What is it?

- [x] Feature / enhancement
- [ ] Bug
- [ ] Docs / tests / types / typos

# Description

The current implementation of `zod$` used type casting. I have fixed the types and solved the problems that led to the current type-casting solution. I also added support for my library Valibot with a `valibot$` validator that works similarly to the `zod$` validator.

# Checklist:

- [x] My code follows the [developer guidelines of this project](https://github.com/QwikDev/qwik/blob/main/CONTRIBUTING.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] Added new tests to cover the fix / functionality

### CODE REVIEW

**Comment by fabian-hiller** [comment]
This is a draft PR because it is missing docs and probably a few other things. I prefer to get a review and confirmation that this will be merged before finalizing this PR.

**Comment by wmertens** [comment]
Can you run `pnpm api.update` so we can see what the changes are?

**Comment by wmertens** [comment]
LGTM - we could add it as an alpha preview feature in v1.8 IMHO, saying that we're looking for feedback

**Comment by fabian-hiller** [comment]
Thank you for your review!

> Can you run `pnpm api.update` so we can see what the changes are?

Will do this tomorrow.

> we could add it as an alpha preview feature in v1.8 IMHO, saying that we're looking for feedback

Not sure if this is necessary but we can do it.

**Comment by fabian-hiller** [comment]
How quickly can this be merged? Should I continue with the documentation?

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (wmertens) [inline packages/qwik-city/src/runtime/src/server-functions.ts:244]: am I reading this correctly that it won't tell you which array item has the issue? Is that not possible to do?

**Review by wmertens — COMMENTED** [review]
I think we can merge this as a new feature provided we're sure that the error output will not change later.

Then we'd have to have some tests that verify the error shape.

**Comment by wmertens** [comment]
I think it straightforward enough to merge asap, it could be part of 1.8.0. Documentation would be great, please add some tests though, see https://github.com/QwikDev/qwik/blob/main/packages/qwik/src/core/render/jsx/types/jsx-types.unit.tsx for examples on how to test types, as well as real tests for the error shape.
Could you also `pnpm change add` to add a changeset?

File: packages/qwik-city/package.json
Comment (fabian-hiller) [reply packages/qwik-city/package.json:12]: I used a fixed version number because the same approach was taken with Zod. As long as there are no breaking changes, it doesn't matter. So we can change it to `"^0.33.0"` or `">=0.33.0 <1"` and update to v1 when it is available. Please let me know how to proceed.

Comment (wmertens) [reply packages/qwik-city/package.json:12]: With zod it actually needs the fixed version because the minor update breaks the types :( With valibot I hope that won't happen :)

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: That's true and I think it's bad, but I just followed the Zod implementation because changing it would lead to breaking changes due to the `ValidatorErrorKeyDotNotation` type.

**Comment by fabian-hiller** [comment]
Will do. Is it true that the Zod implementation is not tested, or am I just not finding the tests?

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (wmertens) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: Hmm, could we not have a different error signature for valibot? Make it a generic type?

**Comment by wmertens** [comment]
> Will do. Is it true that the Zod implementation is not tested, or am I just not finding the tests?

indeed no tests currently, but added in #6720

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: We could do this, but it would lead to a major rewrite of the types in `types.ts`. Also, switching from `zod$` to `valibot$` gets confusing as the error implementation is inconsistent. I have heard that there are plans to rewrite Qwik City. This is something we could tackle as part of that.

Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: I recommend choosing a unified format without edge cases, or making it custom and just building on top of the schema library's error formatting. Both Zod and Valibot contain a `flatten` function/method.

**Comment by fabian-hiller** [comment]
@wmertens does my last `pnpm change add` commit looks good to you or is something missing?

File: packages/qwik-city/package.json
Comment (fabian-hiller) [reply packages/qwik-city/package.json:12]: I don't expect breaking changes. Should I change it to `>=0.33.0 <2`?

Comment (wmertens) [reply packages/qwik-city/package.json:12]: Yes please

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (wmertens) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: I prefer a per-validator format. That way if people want to roll their own they can, and valibot can already provide nicer errors.

**Comment by wmertens** [comment]
> @wmertens does my last `pnpm change add` commit looks good to you or is something missing?

Could you write it as a report of things changed instead? So not "I changed X to Y" but instead "X is now Y".

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: I agree, but how should I proceed with this PR? I don't think we should break the current `zod$` implementation. Also, it might be complicated to change this only for Valibot because we probably need to change/break a few types.

Comment (wmertens) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: But if each validator has its own return type, we're not breaking zod$ are we?

Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: The current `zod$` implementation used generic names for types like `GetValidatorType` and `TypedDataValidator`. We could of course add `GetValibotValidatorType` and `TypedValibotDataValidator` and add new overload signatures only for Valibot, but that would make the code even more complicated.

Since `zod$` does not use a Zod-specific return type, I just "plugged" the `valibot$` implementation into the `zod$` one and improved a few things. It might be easier to stick with my current solution for now and rewrite it at some point.

Comment (wmertens) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: But if we do it now, it's a new feature, and if we do it later, it's a breaking feature, no?

Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: Yes, that's true, but we'll probably break it for `zod$` at some point anyway.

Comment (wmertens) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: If we have this generic per-validator API we wouldn't have to, we'd just make zod2$.

Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: That's true, but I'm not sure we should increase the number for each breaking change. In the end, I'm fine with either solution. So feel free to make a final decision, and I will try my best to implement it.

Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: Another breaking change we should probably consider in the long run is to remove the export of Zod's `z`. I am not sure if such a library should be "baked" into Qwik City. It's probably better to just use the version of Zod that the user has deliberately installed. But that's another topic.

**Comment by tzdesign** [comment]
@wmertens @fabian-hiller is there something I can help with?

I need this isAny check in the types to upgrade qwik. We still hang on 1.5.6 with this neat JSX security issue.

**Comment by fabian-hiller** [comment]
Feel free to take a look at the currently unresolved error shape implementation and share your thoughts on how to proceed: https://github.com/QwikDev/qwik/pull/6752#pullrequestreview-2215059841

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (maiieul) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: If we want to strictly follow semver we can create zod2$ and in V2 deprecate zod2$ and use zod2$'s code for zod$. That way we don't have to do a breaking change for valibot$ and it can start to share code with zod2$ (+ the benefits for users of zod2$) right now.

But if it's just easier for now to go with zod$ and valibot$ that's good too. Valibot users are quite bleeding-edgy so they shouldn't complain if there's a breaking change for valibot$ in the future.

I think you can take the decision @fabian-hiller since you have more context. Both are fine imo. Easier and cleaner code might be more important than another optimization atm.

> Another breaking change we should probably consider in the long run is to remove the export of Zod's z
Would be good to have those zod breaking changes in the same release. Could you create an issue so that we don't forget?

Comment (brandonpittman) [reply packages/qwik-city/src/runtime/src/server-functions.ts:244]: > But if it's just easier for now to go with zod$ and valibot$ that's good too. Valibot users are quite bleeding-edgy so they shouldn't complain if there's a breaking change for valibot$ in the future.

Let’s not adopt a  breaking change is okay mentality from the get-go. The recent Valibot API change was a not-so-small problem in our apps at work recently.

**Comment by octet-stream** [comment]
Looks like this PR fixes type inference for zod validator, and if so, then it will solve this issue: https://github.com/QwikDev/qwik/issues/5784

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (tzdesign) [inline packages/qwik-city/src/runtime/src/server-functions.ts:240]: @fabian-hiller I don't know much about valibot, what is that key? Is it a wrong valibot version, that it fails?

Sorry to get back to you guys so late. We had 10 hour shifts fixing the qwik-shop.

**Review by tzdesign — CHANGES_REQUESTED** [review]
@fabian-hiller can you fix the key thing, I would than check it out in our large code base again.

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:240]: Which version of Valibot do you use? We will hit v1 RC with a stable API soon.

**Comment by fabian-hiller** [comment]
I will try to go over everything in the next few days and share my thoughts on how to proceed.

File: packages/qwik-city/src/runtime/src/server-functions.ts
Comment (tzdesign) [reply packages/qwik-city/src/runtime/src/server-functions.ts:240]: @fabian-hiller none. This might be the problem?

Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:240]: I am not an expert on the Qwik repo and cannot help with that. I don't think this "problem" is related to this PR.

Comment (fabian-hiller) [reply packages/qwik-city/src/runtime/src/server-functions.ts:240]: Okay. I think I know what you mean now. I will take a look.

**Comment by fabian-hiller** [comment]
@wmertens I have thought about this PR and think it is best if we mark the new `valibot$` adapter as "alpha" and merge it as is. The main reason is that this PR improves many types and fixes some problems with the current `zod$` implementation without introducing breaking changes. If we were to start rewriting everything to support the default error format of the validator being used, this PR would probably never get merged as I have less time at the moment. After merging, someone else could change the error format and figure out how to rewrite the types.

**Comment by fabian-hiller** [comment]
@wmertens I am not sure what to do with the tests. This is because the type `ValidatorErrorType` does not match the implementation of `flattenZodIssues` in edge cases if I understand the code correctly. Before investing a day of work in fixing the types and writing meaningful tests, I would rather not write tests, merge this PR as is and create a new PR for Qwik City v2 that fixes and changes the implementation of `zod$` and `valibot$` by using the native error format of the validation library. Otherwise, we will be forced to write the tests twice.

**Comment by fabian-hiller** [comment]
My current recommendation is to review and merge this PR as is. I plan to create another PR for Qwik City v2 that will change the error formatting to the native output, further improve the types, and add unit and type tests.

**Review by wmertens — APPROVED** [review]
:+1: :cowboy_hat_face:  :+1:

**Comment by brandonpittman** [comment]
Happy to report this is working as I had hoped it would last year. Thank you @fabian-hiller for seeing it over the finish line! 🥳

**Comment by brandonpittman** [comment]
This is working really well. When's it going to be out of _experimental_?

**Comment by fabian-hiller** [comment]
I am not sure it will ever be. Maybe a `schema$` function build on top of the Standard Schema interface I am working on with Colin, David and others is the way to go for Qwik.

File: packages/qwik-city/package.json
Comment (shairez) [reply packages/qwik-city/package.json]: @fabian-hiller @wmertens apparently there were some changes done for 1.0.0-alpha that caused one of my projects to fail using `globalAction$` because it was upgraded from an older qwik-city version but kept using the older version because it was in the range.

I think we need to get to the standard-schema solution asap to make zod / valibot optional

**Pull request merged**
